doc/bgp_error_handling/BGP_Route_Error_Handling_Arlo.md

```diff
@@ -0,0 +1,241 @@
+
+
+  
+# BGP Route Install Error Handling
+# High Level Design Document
+#### Rev 0.1
+
+# Table of Contents
+  * [List of Tables](#list-of-tables)
+  * [Revision](#revision)
+  * [About This Manual](#about-this-manual)
+  * [Scope](#scope)
+  * [Definition/Abbreviation](#definitionabbreviation)
+  
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+# Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 05/07/2019  |   Sudhanshu Kumar  | Initial version                   |
+
+# About this Manual
+This document provides information about how to handle the "route add failure in hardware" related errors in BGP in SONIC.
+# Scope
+This document describes the high level design of BGP route install error handling feature. Implementation for warm reboot and GR for BGP is out of scope for this feature. When route installation fails in hardware due to table full, BGP may retry again when some routes get deleted. This Retry mechanism in BGP for failed routes will not be implemented in this release.
+
+# Definition/Abbreviation
+### Table 1: Abbreviations
+
+| **Term**          | ***Meaning***           |
+|-------------------|-------------------------|
+| BGP               | Border Gateway Protocol |
+| GR                | Graceful Restart        |
+| SONIC             | Software for Open Networking in the Cloud                        |
+| FRR    	        |  FRRouting                       |
+| FPM               |  Forwarding Plane Manager        |
+| SwSS              |  SONiC Switch State Service      |
+# 1 Requirement Overview
+ When BGP learns a prefix, it sends it to route table manager(Zebra). The routes are installed in kernel and sent to APP_DB via fpmsyncd. 
+The Orchagent reads the route from APP_DB, creates new resources like nexthop or nexthop group Id and installs the route in ASIC_DB. The syncd triggers the appropriate SAI API and route is installed in hardware. The CRM manages the count of critical resources allocated by orchagent through SAI API.
+Due to resource allocations failures in hardware, SAI API calls can fail and these failures should be notified to Zebra and BGP.
+On learning the prefix, BGP can immediately advertise the prefix to its neighbors. However, if the error-handling feature is enabled, BGP waits for success notification from hardware before advertising the same to its peers. If the hardware returns error, the routes are not advertised to the peers. 
+
+## 1.1 Functional Requirements
+
+ 
+
+ 1. BGP should not advertise the routes which have failed to be installed in hardware.
+ 1. BGP should mark the routes which are not installed in hardware as  "FIB-install pending" routes in its RIB-IN table.

```

Comment:
what is the reconciliation on process/daemon crashes/restarts? how to reconcile the routes? Please list down the scenarios and the observations?

Reply:
There are 2 scenarios here. One is warm-reboot case and another is unplanned reboot (like docker restart due to cold reboot). 
During warm reboot, fpmsyncd supports syncing of existing routes in APP_DB with newly learnt routes.  After warm reboot of BGP docker, BGP sends newly learnt routes to fpmsyncd. Since warm reboot is enabled, fpmsyncd will mark the old routes and send only the newly learnt routes to APP_DB. In our case, since we do not store any error route in APP_DB, so we 
don't need to do anything here.
For e.g., before warm reboot, suppose, we had 5 routes sent by BGP to fpmsynd, out of them., 5th route failed to be installed in hardware. Zebra will delete the 5th route from kernel, APP_DB, ASIC_DB and ERROR_DB. After warm reboot, when fpmsyncd receives the 5 routes again from BGP, it will send add for the 5th route again (since 5th route was never present in APP_DB prior to warm reboot).
During unplanned reboot, for the same 5 routes, suppose, fpmsyncd crashed before processing the add failure for the 5th route. Now, all the 5 routes are present in APP_DB and ASIC_DB. Since 
the 5th route failed to be installed in hardware, it is present in ERROR_DB. When docker comes up again, if warm-reboot feature is not enabled, all 5 routes will be pushed by fpmsyncd.
If warm-reboot feature for BGP is enabled, fpmsynd will reconcile the 5 routes with those present in APP_DB. In this case, it will receive the notification for the 5th route failure from ERROR_DB. Or, fpmsyncd, during
warm-reboot should check if any add failure entry is present in ERROR_DB and should process it by sending delete to the APP_DB.